From cdbb58b28212b24f77629e3eddd84cf2d71857e1 Mon Sep 17 00:00:00 2001
From: Bart Ribbers <bribbers@disroot.org>
Date: Mon, 20 Dec 2021 22:29:46 +0100
Subject: [PATCH] Add CEC daemon script

Allows input using the TV remote on CEC supported devices.
Adapted and modernized from https://invent.kde.org/plasma-bigscreen/bigscreen-debos-image-rpi4/-/blob/master/overlays/bigscreen/etc/skel/.config/autostart-scripts/cec-daemon.py

Tested on Wayland and libcec >= 6.0.x
---
 CMakeLists.txt               |  2 +
 bin/40-uinput.rules          |  1 +
 bin/CMakeLists.txt           |  6 +++
 bin/cec-daemon.desktop.cmake |  5 ++
 bin/cec-daemon.py            | 99 ++++++++++++++++++++++++++++++++++++
 5 files changed, 113 insertions(+)
 create mode 100644 bin/40-uinput.rules
 create mode 100644 bin/cec-daemon.desktop.cmake
 create mode 100755 bin/cec-daemon.py

diff --git a/CMakeLists.txt b/CMakeLists.txt
index ecdf04e..8ad1961 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -64,6 +64,8 @@ plasma_install_package(lookandfeel org.kde.plasma.mycroft.bigscreen look-and-fee
 
 
 install(FILES bin/mycroft-skill-launcher.py DESTINATION ${BIN_INSTALL_DIR})
+install(FILES bin/cec-daemon.py DESTINATION ${BIN_INSTALL_DIR})
+install(FILES bin/40-uinput.rules DESTINATION ${LIB_INSTALL_DIR}/udev/rules.d)
 
 if(ECM_VERSION VERSION_GREATER_EQUAL 5.79)
     include(KDEClangFormat)
diff --git a/bin/40-uinput.rules b/bin/40-uinput.rules
new file mode 100644
index 0000000..918e0a7
--- /dev/null
+++ b/bin/40-uinput.rules
@@ -0,0 +1 @@
+SUBSYSTEM=="misc", KERNEL=="uinput", MODE="0660", GROUP="video"
diff --git a/bin/CMakeLists.txt b/bin/CMakeLists.txt
index 2f1bb3b..85eee3b 100644
--- a/bin/CMakeLists.txt
+++ b/bin/CMakeLists.txt
@@ -1,6 +1,12 @@
 configure_file(plasma-bigscreen-wayland.in ${CMAKE_CURRENT_BINARY_DIR}/plasma-bigscreen-wayland)
 install(PROGRAMS ${CMAKE_CURRENT_BINARY_DIR}/plasma-bigscreen-wayland DESTINATION ${BIN_INSTALL_DIR})
 
+configure_file(cec-daemon.desktop.cmake ${CMAKE_CURRENT_BINARY_DIR}/cec-daemon.desktop)
+install(FILES
+	${CMAKE_CURRENT_BINARY_DIR}/cec-daemon.desktop
+	DESTINATION ${KDE_INSTALL_SYSCONFDIR}/xdg/autostart
+)
+
 configure_file(plasma-bigscreen-wayland.desktop.cmake ${CMAKE_CURRENT_BINARY_DIR}/plasma-bigscreen-wayland.desktop)
 install(FILES
         ${CMAKE_CURRENT_BINARY_DIR}/plasma-bigscreen-wayland.desktop
diff --git a/bin/cec-daemon.desktop.cmake b/bin/cec-daemon.desktop.cmake
new file mode 100644
index 0000000..00a47a8
--- /dev/null
+++ b/bin/cec-daemon.desktop.cmake
@@ -0,0 +1,5 @@
+[Desktop Entry]
+Exec=cec-daemon.py
+Name=CEC Daemon
+Description=Allows input using TV remote controls through the CEC protocol
+X-KDE-PluginInfo-Version=${PROJECT_VERSION}
diff --git a/bin/cec-daemon.py b/bin/cec-daemon.py
new file mode 100755
index 0000000..e9f6f2d
--- /dev/null
+++ b/bin/cec-daemon.py
@@ -0,0 +1,99 @@
+#!/usr/bin/python3
+
+import cec
+import uinput
+import psutil
+
+u = uinput
+
+KEYMAP = {
+   0: u.KEY_ENTER,
+   1: u.KEY_UP,
+   2: u.KEY_DOWN,
+   3: u.KEY_LEFT,
+   4: u.KEY_RIGHT,
+   9: u.KEY_HOMEPAGE,
+   10: u.KEY_MENU,
+   13: u.KEY_BACK,
+   44: u.KEY_HOMEPAGE,
+   68: u.KEY_PLAY,
+   69: u.KEY_STOP,
+   70: u.KEY_PAUSE,
+   75: u.KEY_FASTFORWARD,
+   76: u.KEY_REWIND,
+   103: u.KEY_HOMEPAGE,
+   113: u.KEY_BLUE,
+   114: u.KEY_RED,
+   115: u.KEY_YELLOW,
+   116: u.KEY_GREEN
+}
+
+device = uinput.Device(KEYMAP.values())
+
+keystate = None
+override = False
+
+
+def check_override():
+    global override
+    if "kodi-x11" or "kodi-wayland" in (
+            p.name() for p in psutil.process_iter()):
+        override = True
+    else:
+        override = False
+
+
+def init_cec():
+    cecconfig = cec.libcec_configuration()
+    cecconfig.strDeviceName = "libcec"
+    cecconfig.bActivateSource = 0
+    cecconfig.deviceTypes.Add(cec.CEC_DEVICE_TYPE_RECORDING_DEVICE)
+    cecconfig.clientVersion = cec.LIBCEC_VERSION_CURRENT
+
+    cecconfig.SetKeyPressCallback(keypress_callback)
+
+    lib = cec.ICECAdapter.Create(cecconfig)
+
+    # Wait till we get a CEC device
+    print("Detecting CEC adapters...")
+    returned_adapter = None
+    while returned_adapter is None:
+        adapters = lib.DetectAdapters()
+        for adapter in adapters:
+            print("Found a CEC adapter:")
+            print("Port:    " + adapter.strComName)
+            print("Vendor:  " + hex(adapter.iVendorId))
+            print("Product: " + hex(adapter.iProductId))
+            returned_adapter = adapter.strComName
+
+    try:
+        lib.Open(returned_adapter)
+        print("Connection opened")
+
+        main_loop()
+    except Exception:
+        print("Failure while connecting to the adapter")
+
+
+def main_loop():
+    while True:
+        # Just make sure we live forever
+        pass
+
+
+def keypress_callback(key, duration):
+    global keystate
+
+    if override:
+        return
+
+    if keystate is None:
+        device.emit(KEYMAP[key], 1)
+        keystate = "down"
+    else:
+        device.emit(KEYMAP[key], 0)
+        keystate = None
+
+
+if __name__ == '__main__':
+    init_cec()
-- 
GitLab

